server/schedule/placement/rule.go

```diff
@@ -54,26 +54,46 @@ type Rule struct {
 	LabelConstraints []LabelConstraint `json:"label_constraints,omitempty"` // used to select stores to place peers
 	LocationLabels   []string          `json:"location_labels,omitempty"`   // used to make peers isolated physically
 	IsolationLevel   string            `json:"isolation_level,omitempty"`   // used to isolate replicas explicitly and forcibly
+
+	group *RuleGroup // only set at runtime, no need to {,un}marshal or persist.
 }
 
-func (r Rule) String() string {
+func (r *Rule) String() string {
 	b, _ := json.Marshal(r)
 	return string(b)
 }
 
 // Key returns (groupID, ID) as the global unique key of a rule.
-func (r Rule) Key() [2]string {
+func (r *Rule) Key() [2]string {
 	return [2]string{r.GroupID, r.ID}
 }
 
 // StoreKey returns the rule's key for persistent store.
-func (r Rule) StoreKey() string {
+func (r *Rule) StoreKey() string {
 	return hex.EncodeToString([]byte(r.GroupID)) + "-" + hex.EncodeToString([]byte(r.ID))
 }
 
+func (r *Rule) groupIndex() int {
+	if r.group != nil {
+		return r.group.Index
+	}
+	return 0
+}
+
+// RuleGroup defines properties of a rule group.
+type RuleGroup struct {
+	ID       string `json:"id,omitempty"`
```

Comment:
Does this field conflict with `Rule. GroupID`?

Reply:
Rules and RuleGroups will be maintained and stored separately. ID is used to associate them together. (consider it as a foreign key)

Reply:
I got it.